File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgViewPagerViewHolder.java
package org.tvheadend.tvhclient.ui.features.epg;

import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.domain.entity.EpgChannel;
import org.tvheadend.tvhclient.domain.entity.EpgProgram;

import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;

import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import timber.log.Timber;

public class EpgViewPagerViewHolder extends RecyclerView.ViewHolder {

    private final EpgProgramListRecyclerViewAdapter recyclerViewAdapter;
    private final FragmentActivity activity;
    private final long startTime;
    private final long endTime;
    private final EpgViewModel viewModel;
    private final ScheduledExecutorService execService = Executors.newScheduledThreadPool(10);

    @BindView(R.id.program_list_recycler_view)
    protected RecyclerView recyclerView;
    @BindView(R.id.progress_bar)
    ProgressBar progressBar;
    @BindView(R.id.no_programs)
    TextView noProgramsTextView;

    EpgViewPagerViewHolder(FragmentActivity activity, View view, float pixelsPerMinute, long startTime, long endTime, RecyclerView.RecycledViewPool viewPool) {
        super(view);
        ButterKnife.bind(this, view);

        this.activity = activity;
        this.startTime = startTime;
        this.endTime = endTime;

        recyclerView.setLayoutManager(new CustomHorizontalLayoutManager(view.getContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(view.getContext(), LinearLayoutManager.HORIZONTAL));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setRecycledViewPool(viewPool);
        recyclerViewAdapter = new EpgProgramListRecyclerViewAdapter(pixelsPerMinute, startTime, endTime);
        recyclerView.setAdapter(recyclerViewAdapter);

        viewModel = ViewModelProviders.of(activity).get(EpgViewModel.class);
    }

    void bindData(final EpgChannel epgChannel) {

        recyclerView.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);
        noProgramsTextView.setVisibility(View.GONE);

        execService.execute(() -> {
            List<EpgProgram> programs = viewModel.getProgramsByChannelAndBetweenTimeSync(epgChannel.getId(), startTime, endTime);
            if (programs.size() > 0) {
                Timber.d("Loaded " + programs.size() + " programs for channel " + epgChannel.getName());
                activity.runOnUiThread(() -> {
                    recyclerViewAdapter.addItems(programs);
                    recyclerView.setVisibility(View.VISIBLE);
                    progressBar.setVisibility(View.GONE);
                    noProgramsTextView.setVisibility(View.GONE);
                });
            } else {
                Timber.d("Loaded no programs for channel " + epgChannel.getName());
                activity.runOnUiThread(() -> {
                    recyclerView.setVisibility(View.GONE);
                    progressBar.setVisibility(View.GONE);
                    noProgramsTextView.setVisibility(View.VISIBLE);
                });
            }
        });

        viewModel.getRecordingsByChannel(epgChannel.getId()).observe(activity, recordings -> {
            if (recordings != null) {
                recyclerViewAdapter.addRecordings(recordings);
            }
        });
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/epg/EpgViewPagerViewHolder.kt
package org.tvheadend.tvhclient.ui.features.epg

import android.view.View
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.epg_program_list_adapter.*
import org.tvheadend.tvhclient.domain.entity.EpgChannel
import org.tvheadend.tvhclient.ui.common.gone
import org.tvheadend.tvhclient.ui.common.visible
import timber.log.Timber
import java.util.concurrent.Executors

class EpgViewPagerViewHolder(override val containerView: View, private val activity: FragmentActivity, private val epgViewModel: EpgViewModel, private val fragmentId: Int, viewPool: RecyclerView.RecycledViewPool) : RecyclerView.ViewHolder(containerView), LayoutContainer {

    private val recyclerViewAdapter: EpgProgramListRecyclerViewAdapter
    private val viewModel: EpgViewModel = ViewModelProviders.of(activity).get(EpgViewModel::class.java)
    private val execService = Executors.newScheduledThreadPool(10)

    init {
        program_list_recycler_view.layoutManager = CustomHorizontalLayoutManager(containerView.context)
        program_list_recycler_view.addItemDecoration(DividerItemDecoration(containerView.context, LinearLayoutManager.HORIZONTAL))
        program_list_recycler_view.itemAnimator = DefaultItemAnimator()
        program_list_recycler_view.setRecycledViewPool(viewPool)
        recyclerViewAdapter = EpgProgramListRecyclerViewAdapter(viewModel, fragmentId)
        program_list_recycler_view.adapter = recyclerViewAdapter
    }

    fun bindData(epgChannel: EpgChannel) {

        program_list_recycler_view.gone()
        progress_bar.visible()
        no_programs.gone()

        execService.execute {
            val programs = viewModel.getProgramsByChannelAndBetweenTimeSync(epgChannel.id, fragmentId)
            if (programs.isNotEmpty()) {
                Timber.d("Loaded ${programs.size} programs for channel ${epgChannel.name}")
                activity.runOnUiThread {
                    recyclerViewAdapter.addItems(programs.toMutableList())
                    program_list_recycler_view.visible()
                    progress_bar.gone()
                    no_programs.gone()
                }
            } else {
                Timber.d("Loaded no programs for channel ${epgChannel.name}")
                activity.runOnUiThread {
                    program_list_recycler_view.gone()
                    progress_bar.gone()
                    no_programs.visible()
                }
            }
        }

        viewModel.getRecordingsByChannel(epgChannel.id).observe(activity, Observer { recordings ->
            if (recordings != null) {
                recyclerViewAdapter.addRecordings(recordings)
            }
        })
    }
}
